pkg/controller/node/nodecontroller.go

```diff
@@ -926,3 +1083,139 @@ func (nc *NodeController) ComputeZoneState(nodeReadyConditions []*api.NodeCondit
+		newTaints := []api.Taint{}
+		for _, oldTaint := range oldTaints {
+			if taintToAdd.MatchTaint(oldTaint) {
+				return []api.Taint{}, false, nil
+			}
+		}
+		newTaints = append(oldTaints, *taintToAdd)
+		return newTaints, true, nil
+	}
+
+	return tryModifyNodeTaints(kubeClient, nodeName, addTaints)
+}
+
+// tryRemoveTaintsOffNode tries to remove taints off a node,
+// return true if taints removed, or return false if taints don't exist.
+func tryRemoveTaintsOffNode(kubeClient clientset.Interface, nodeName string, taintsToRemove ...api.Taint) (bool, error) {
+	removeTaints := func(taints []api.Taint) ([]api.Taint, bool, error) {
+		removed := false
+		for _, taintToRemove := range taintsToRemove {
+			updated := false
+			taints, updated = api.DeleteTaint(taints, taintToRemove)
+			removed = removed || updated
+		}
+
+		return taints, removed, nil
+	}
+
+	return tryModifyNodeTaints(kubeClient, nodeName, removeTaints)
+}
+
+// trySyncNodeTaintsWithNodeStatus will try to update node taint according to node status.
+// If node is observed not ready (NodeReady status ConditionFalse), add taint `notready:NoExecute` to node,
+// if node is observed unknown (NodeReady status ConditionUnknown), add taint `unreachable:NoExecute` to node,
+// if node is observed ready, remove these two taints.
+// TODO: may incorporate these into tryUpdateNodeStatus to reduce api call when taint becomes a field of nodeStatus
+func (nc *NodeController) trySyncNodeTaintsWithNodeStatus(nodeName string, decisionTimestamp unversioned.Time, observedReadyCondition, currentReadyCondition *api.NodeCondition) {
+	if currentReadyCondition == nil {
+		return
+	}
+
+	nodeNotReadyTaint := api.Taint{
+		Key:       unversioned.TaintNodeNotReady,
+		Effect:    api.TaintEffectNoExecute,
+		TimeAdded: nc.now(),
+	}
+
+	nodeUnreachableTaint := api.Taint{
+		Key:       unversioned.TaintNodeUnreachable,
+		Effect:    api.TaintEffectNoExecute,
+		TimeAdded: nc.now(),
+	}
+
+	var taintToAdd *api.Taint
+	var taintsToRemove []api.Taint
+
+	if observedReadyCondition.Status == api.ConditionFalse &&
+		decisionTimestamp.After(nc.nodeStatusMap[nodeName].readyTransitionTimestamp.Add(nc.podEvictionTimeout)) {
+		taintToAdd = &nodeNotReadyTaint
+	}
+	if observedReadyCondition.Status == api.ConditionUnknown &&
+		decisionTimestamp.After(nc.nodeStatusMap[nodeName].probeTimestamp.Add(nc.podEvictionTimeout)) {
+		taintToAdd = &nodeUnreachableTaint
+	}
+	if observedReadyCondition.Status == api.ConditionTrue {
+		taintsToRemove = []api.Taint{
+			{
+				Key:    unversioned.TaintNodeNotReady,
+				Effect: api.TaintEffectNoExecute,
+			},
+			{
+				Key:    unversioned.TaintNodeUnreachable,
+				Effect: api.TaintEffectNoExecute,
+			},
+		}
+	}
+
+	if currentReadyCondition.Status != api.ConditionTrue && observedReadyCondition.Status == api.ConditionTrue {
+		taintToAdd = &nodeUnreachableTaint
+	}
+
+	if taintToAdd != nil {
+		added, err := tryAddTaintToNode(nc.kubeClient, nodeName, taintToAdd)
```

Comment:
After you incorporate this function into 'tryUpdateNodeStatus' Taint updates will go in in the single transaction as NodeReady updates. It's crucial to do that, as otherwise we might end up in a pretty bad state when the NodeReady gets updated, and updating Taint fails.